lib/widgets/message_list.dart

```diff
@@ -530,6 +534,68 @@ class MarkAsReadWidget extends StatelessWidget {
   }
 }
 
+class _TypingStatusState extends State<TypingStatusWidget> with PerAccountStoreAwareStateMixin<TypingStatusWidget> {
+  TypingStatus? model;
+
+  @override
+  void onNewStore() {
+    model?.removeListener(_modelChanged);
+    model = PerAccountStoreWidget.of(context).typingStatus
+      ..addListener(_modelChanged);
+  }
+
+  @override
+  void dispose() {
+    model?.removeListener(_modelChanged);
+    super.dispose();
+  }
+
+  void _modelChanged() {
+    setState(() {
+      // The actual state lives in [model].
+      // This method was called because that just changed.
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final narrow = widget.narrow;
+    const placeholder = SizedBox(height: 8);
+    if (narrow is! SendableNarrow) return placeholder;
+
+    final localization = ZulipLocalizations.of(context);
+    final typistNames = model!.typistIdsInNarrow(narrow)
+      .where((id) => id != store.selfUserId)
+      .map((id) => store.users[id]?.fullName ?? localization.unknownUserName)
+      .toList();
+    if (typistNames.isEmpty) return placeholder;
+
+    final String text = switch (typistNames.length) {
+      1 => localization.onePersonTyping(typistNames[0]),
+      2 => localization.twoPeopleTyping(typistNames[0], typistNames[1]),
+      _ => localization.manyPeopleTyping,
+    };
+
+    return Padding(
+      padding: const EdgeInsetsDirectional.only(start: 16, top: 2),
+      child: Text(text,
+        textAlign: TextAlign.start,
+        style: const TextStyle(
+          color: HslColor(0, 0, 53), fontStyle: FontStyle.italic)),
+    );
+  }
+}
+
+class TypingStatusWidget extends StatefulWidget {
```

Comment:
nit (a thought prompted by looking at the order of the tests): it'd be most logical for this to come just above MarkAsReadWidget, rather than just below — since that matches how it appears in the UI

(That would be the opposite order from their call sites, but that's because the code at the call sites happens to be written in a context where it's counting from the bottom.)